### ISSUE TITLE
S1117 complains about hiding a field in a static method

### ISSUE THREAD

**Issue by git-ruttmann** [SonarSource/sonar-dotnet#3449]
### Description

__This is more a discussion than an issue.__

Instead of using overloaded Constructors with complex logic, I prefer a static builder method, in my case `Polygon.FromVertexList`. The parameters are translated to the internal structures and passed to the constructor. 

I choose that approach because:
* The static method can use the private data types of the class.
* The conversion logic is encapsulated.
* The static method has a useful name, so it's clear when to use it.
* The constructor stays free from logic.
* The class can use the fields as immutable, private and readonly.
* There can be different input data formats using the same constructor.

While building the internal structures, it makes sense to use the same variable names as the fields, so it's clear what is constructed here.

Rule __S1117__ complains about hiding the field. But fields arent't accessible from static methods, so IMHO, this is a false positive.

### Repro steps

``` csharp
public class Polygon
{
  private readonly VertexChain[] vertexChain;
  private Polygon(VertexChain[] vertexChain)
  {
    this.vertexChain = vertexChain;
  }
  
  public static Polygon FromVertexList(IEnumerable<Vector2> vertices)
  {
    // using the same name triggers S1117.
    var vertexChain = vertices
      .Select((v, i) => new VertexChain { Next = i + 1, Prev = i - 1, Vertex = v })
      .ToArrray();
    vertexChain[0].Prev = vertexChain.Length - 1;
    vertexChain[vertexChain.Length - 1].Next = 0;
    return new Polygon(vertexChain);
  }

  private struct VertexChain
  {
    int Next;
    int Prev;
    Vector2 Vertex;
  }
}
```

### Expected behavior

_This is the difficult part._
The rule is definitly great for fields that are really hidden.
But I'd like to disable the rule for static methods using the same variable names as fields.

My suggestion is to split the situation "static method uses member field" in a different rule. That rule can be deactivated as per personal preference.

I'd be happy to get the opinion of others.

### Known workarounds

It's possible to use a private property instead of a private field and start the property name with a capital letter.

### Related information

* SonarLint version 4.23.0.19399 on Visual Studio 2019

**Comment by pavel-mikula-sonarsource** [comment]
Hi @git-ruttmann ,

Your design, reasoning and suggestion look all valid to me. But I'm not sure it's worth adding a parameter. It's a corner case so it cannot be a separate rule and we're trying to keep rules without parameters to avoid parameter hell in configuration. This rule is quite old (low rule ID) and this is the first time such suggestion occurs. So I don't believe many users would use such a parameter.

I would vote for closing this discussion. 

@andrei-epure-sonarsource @costin-zaharia-sonarsource Any thoughts on this topic?

**Comment by andrei-epure-sonarsource** [comment]
My 2 cents: I don't think we need to add a parameter. When the rule raises issues on a static method regarding an instance field, it's indeed an FP, as the instance field cannot be accessed inside the method.

I wouldn't create a different rule either. Just make sure that S1117
- in a non-static method, should not shadow static or non-static fields
- in a static method, do not shadow static fields

**Comment by pavel-mikula-sonarsource** [comment]
@andrei-epure-sonarsource It's easy to miss that a method is static. Once it shadows a class field, you could believe that you're actually modifying the class field in place where you should remove the static modifier. That's why the rule exists IMHO to avoid this kind of confusion when shadowing anything.

That all contrary to this corner case, where it's reasonable to shadow.

**Comment by pavel-mikula-sonarsource** [comment]
Hi @git-ruttmann,

We had a short discussion about this topic and the conclusion was that we're not going to change the rule. Shadowing is generally bad practice and can lead to confusion. That's aligned with S1117 type that's a code smell.

This contrary your case, where the reasoning makes sense for this small class. We suggest to close this issue as False Positive in your scenario.
